components/region_cache_memory_engine/src/engine.rs

```diff
@@ -27,31 +28,156 @@ fn cf_to_id(cf: &str) -> usize {
     }
 }
 
+// todo: remove it
+fn readable_user_key(user_key: &[u8]) -> String {
+    let idx = user_key.len() - ENC_KEY_SEQ_LENGTH;
+    format!(
+        "{:?}{}",
+        String::from_utf8(user_key[..idx].to_vec()).unwrap(),
+        !u64::from_be_bytes(user_key[idx..idx + ENC_KEY_SEQ_LENGTH].try_into().unwrap())
+    )
+}
+
+#[derive(Debug, Clone, Copy, PartialEq)]
+enum ValueType {
+    Deletion,
+    Value,
+}
+
+// See `compare` of InternalKeyComparator, for the same user key and same
+// sequence number, ValueType::Value is less than ValueType::Deletion
+const VALUE_TYPE_FOR_SEEK: ValueType = ValueType::Value;
+const VALUE_TYPE_FOR_SEEK_FOR_PREV: ValueType = ValueType::Deletion;
+
+impl ValueType {
+    fn encode(&self) -> u8 {
+        match self {
+            ValueType::Deletion => 0,
+            ValueType::Value => 1,
+        }
+    }
+
+    fn decode(i: u8) -> Self {
+        match i {
+            0 => ValueType::Deletion,
+            1 => ValueType::Value,
+            _ => panic!("invalid"),
+        }
+    }
+}
+
+pub struct InternalKey<'a> {
```

Comment:
Not sure about this myself, but wondering if it may make more sense to use a Bytes in InternalKey. SkipList uses Bytes internally.,Bytes is thread-safe,  avoid copies, but the cost of cloning a Bytes is that of incrementing an Atomic Pointer.  On the other hand, using a reference with a lifetime seems to work so far, and avoids an increment. Just something to keep in mind, potentialy.